### PR TITLE
ARROW-4847: [Python] Add pyarrow.table factory function

### PR DESCRIPTION
**Pull request by jorisvandenbossche**
Rudimentary implementation for https://issues.apache.org/jira/browse/ARROW-4847.

For now I made the minimal choices:

* Which formats to support? For now I only included DataFrame and dictionary, and not list of arrays or list of batches, because DataFrame/dictionary are most unambiguous and don't require additional keywords to be interpreted (eg list of arrays needs a list of names (or the schema)). 
  `from_batches` could probably easily be supported (eg by checking the first element of the list being a RecordBatch), but then it might also be nice to support a list of DataFrames to be interpreted as a list of batches? 

* Which keywords to expose? For now I only included the keywords that are common to the different constructors being dispatched to (in practice only `schema`), in the idea you can always use the specialized constructors if needing more control.

### CODE REVIEW

**Comment by wesm** [comment]
I would say to make `pa.table([a0, a1, a2], columns=['foo', 'bar', 'baz'])` (or `names`?) also work would be reasonable. `Table.from_pydict` does not have a names/columns argument (only schema), so this might be a good time to fix that

**Comment by jorisvandenbossche** [comment]
> I would say to make pa.table([a0, a1, a2], columns=['foo', 'bar', 'baz']) (or names?) also work would be reasonable

pandas is treating such list of arrays as rows:

```
In [11]: pd.DataFrame([np.array([1, 2, 3]), np.array([2, 3, 4])], columns=['a', 'b', 'c']) 
Out[11]: 
   a  b  c
0  1  2  3
1  2  3  4
```

which is not to say that is better / we can't deviate from it, but IMO it certainly indicates that it might not be obvious what the behaviour will be.

> Table.from_pydict does not have a names/columns argument (only schema), so this might be a good time to fix that

What would that argument do? (since the dict already has names in it) To select a subset from the dictionary, or to provide a specific order of the columns? 
It might also be interpreted as overwriting the names from the dict (which I would not do, as that would rely on the order of the dict). Although you could say the same for the `schema` argument of course.

**Comment by jorisvandenbossche** [comment]
> Although you could say the same for the schema argument of course.

Good that I check. This is actually buggy. It is "kind of" overwriting the names of the dictionary, but without properly doing type conversion:

```
In [15]: table = pa.Table.from_pydict(
    ...:     {'a': [1, 2, 3], 'b': [3, 4, 5]},
    ...:     schema=pa.schema([('a', pa.int64()), ('c', pa.int32())]))
In [16]: table                                                                                                                                                                                                     
Out[16]: 
pyarrow.Table
a: int64
c: int32

In [17]: table.to_pandas()                                                                                                                                                                                         
Out[17]: 
   a  c
0  1  3
1  2  0
2  3  4
```

So I am passing a schema with different names, but also with a non-default type (int32 vs int64), and the values in the column are probably first converted as int64 but then viewed as int32 instead of casted (or rather, instead of being converted directly to int32 in the first place).

Will open a JIRA for this.

File: python/pyarrow/table.pxi [hunk 1]
```diff
@@ -1570,6 +1570,35 @@ def _reconstruct_table(arrays, schema):
...
+    See Also
+    --------
+    Table.from_pandas, Table.from_pydict
+    """
+    if isinstance(data, dict):
+        return Table.from_pydict(data, schema=schema)
+    elif isinstance(data, _pandas_api.pd.DataFrame):
+        return Table.from_pandas(data, schema=schema)
+    else:
+        return ValueError("Expected pandas DataFrame or python dictionary")
```
Comment (pitrou) [inline python/pyarrow/table.pxi]: `TypeError` preferably.

**Review by pitrou — COMMENTED** [review]
I think this is a good addition. Just a nit.

File: python/pyarrow/table.pxi (hunk 1 above)
Comment (wesm) [reply python/pyarrow/table.pxi]: done

**Comment by wesm** [comment]
+1. @jorisvandenbossche if you would please use `git rebase` instead of `git merge` in this project, it eases maintenance for us since we are rebase-centric here.

I'll open a follow up JIRA about adding support for `Table.from_arrays` to this function

**Pull request closed without merging**